ambry-api/src/main/java/com/github/ambry/quota/Chargeable.java

```diff
@@ -0,0 +1,39 @@
+/**
+ * Copyright 2021 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.quota;
+
+/**
+ * A {@link Chargeable} is a operation that can be charged against quota.
+ */
+public interface Chargeable {
```

Comment:
I am not sure about the nature of this interface and how it will be used in operation controller?

The `getQuotaResource` would return account id or account/container id, which will be used to charge against a particular quota. But `checkAndCharge` doesn't take `QuotaManager` as its parameter, how would this method check and charge?

BTW, shouldn't "quotaExceedAllowed` be a configuration?

Reply:
This interface will be used by frontend to server operations -- like PutChunk, GetChunk etc. Basically the `OperationController` will maintain a collection of chunk requests that need to be sent to server for each account(or quotaresource). The `OperationController` will send these requests to server when the requests are within quota.
What the `OperationController` gets from each Operation (like PutOperation or GetOperation) are instances of `RequestInfo` classes. The `Chargeable` will be part of this `RequestInfo` class. This will help to make calls like `requestInfo.getChargeable().checkAndCharge()` to check if this chunk request is within quota and if yes then charge for it. If no, then this chunk will be delayed for processing. Note that charging in this case will not be as simple as just saying QuotaManager.charge(). The reason is that for the same chunk, there will be multiple requests sent to server nodes (due to local replica request, remote replica requests, router parallelism etc). But the charging should happen for each chunk only once. In order to manage all this complexity, each of the chunks (like PutChunk) will implement chargeable, and will keep track when and how to charge for the quota.

Reply:
Do note that this interface will be used by OperationController, but they will be implemented by individual chunk classes (like PutChunk).

Reply:
The `getQuotaResource` method will just help OperationController maintain a list of server requests for each `QuotaResource`. The `OperationController` doesn't really use it for charging quota. The problem `getQuotaResource` method in this interface solve is to associate the `RequestInfo` (that represents requests to server nodes) with the account or container to which the request belongs. This will help OperationController maintain a map of pending requests for each account or container.

Reply:
checkAndCharge doesn't need to take QuotaManager as its input. That part will be handled in the implementation of checkAndCharge inside each chunk object.

Reply:
I mean at the end of the day, `checkAndCharge` would need `QuotaManager` to actually charge the quota from the account. So the `QuotaManager` would have to pass to GetOperation and PutOperation.  Right now the quota is charged through a QuotaChargeCallback. Are we going to use the same callback in the `checkAndCharge` implementations?